File: src/so/aporia/utils/user/render/font/FontAtlas.java
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.utils.user.render.font;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import net.minecraft.client.Minecraft;
import net.minecraft.resources.Identifier;
import net.minecraft.server.packs.resources.Resource;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.InputStream;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * Loads and holds glyph data from an MSDF font atlas JSON + texture pair.
 * <p>
 * Загружает и хранит данные глифов из MSDF шрифта (JSON + текстура).
 */
public class FontAtlas {

    private static final Logger LOGGER = LoggerFactory.getLogger("aporia/Font");

    private final Identifier jsonId;
    private final Identifier textureId;
    private final Map<Integer, Glyph> glyphs = new HashMap<>();
    private final AtomicBoolean loaded = new AtomicBoolean(false);

    private float atlasWidth    = 512;
    private float atlasHeight   = 512;
    private float fontSize      = 32;
    private float lineHeight    = 40;
    private float distanceRange = 4;
    private float ascender      = 0.95f;
    private boolean yOriginBottom = false;

    public FontAtlas(Identifier jsonId, Identifier textureId) {
        this.jsonId    = jsonId;
        this.textureId = textureId;
    }

    /**
     * Loads the atlas if not already loaded (thread-safe).
     * <p>
     * Загружает атлас если ещё не загружен (потокобезопасно).
     */
    public void ensureLoaded() {
        if (loaded.get()) return;
        synchronized (this) {
            if (loaded.get()) return;
            doLoad();
        }
    }

    private void doLoad() {
        try {
            Optional<Resource> res = Minecraft.getInstance().getResourceManager().getResource(jsonId);
            if (res.isEmpty()) {
                LOGGER.warn("Font JSON not found in ResourceManager: {}, trying AssetManager...", jsonId);
                loadFromAssetManager();
                return;
            }
            try (InputStream is = res.get().open();
                 InputStreamReader reader = new InputStreamReader(is, StandardCharsets.UTF_8)) {
                parseJson(JsonParser.parseReader(reader).getAsJsonObject());
                loaded.set(true);
                LOGGER.info("Loaded font: {} ({} glyphs)", jsonId, glyphs.size());
            }
        } catch (Exception e) {
            LOGGER.error("Failed to load font: {}", jsonId, e);
            loaded.set(true);
        }
    }
    
    private void loadFromAssetManager() {
        try {
            String path = jsonId.getPath().replace("aporia:", "").replace(".json", "") + ".json";
            java.nio.file.Path assetFile = so.aporia.utils.files.FilesManager.ROOT.resolve(".assets").resolve(path);
            byte[] data = java.nio.file.Files.readAllBytes(assetFile);
            String json = new String(data, StandardCharsets.UTF_8);
            parseJson(JsonParser.parseString(json).getAsJsonObject());
            loaded.set(true);
            LOGGER.info("Loaded font from AssetManager: {} ({} glyphs)", jsonId, glyphs.size());
        } catch (Exception e) {
            LOGGER.error("Failed to load font from AssetManager: {}", jsonId, e);
            loaded.set(true);
        }
    }

    private void parseJson(JsonObject root) {
        if (root.has("atlas")) {
            JsonObject atlas = root.getAsJsonObject("atlas");
            atlasWidth      = getFloat(atlas, "width",         512);
            atlasHeight     = getFloat(atlas, "height",        512);
            fontSize        = getFloat(atlas, "size",           32);
            distanceRange   = getFloat(atlas, "distanceRange",   4);
            if (atlas.has("yOrigin")) {
                yOriginBottom = atlas.get("yOrigin").getAsString().equalsIgnoreCase("bottom");
            }
        }

        if (root.has("metrics")) {
            JsonObject metrics = root.getAsJsonObject("metrics");
            lineHeight = getFloat(metrics, "lineHeight", 1.2f) * fontSize;
            ascender   = getFloat(metrics, "ascender",   0.95f);
        }

        if (root.has("glyphs")) {
            for (JsonElement elem : root.getAsJsonArray("glyphs")) {
                parseGlyph(elem.getAsJsonObject());
            }
        }
    }

    private void parseGlyph(JsonObject g) {
        int unicode = -1;
        if (g.has("unicode"))     unicode = g.get("unicode").getAsInt();
        else if (g.has("char"))   { String s = g.get("char").getAsString(); if (!s.isEmpty()) unicode = s.codePointAt(0); }
        else if (g.has("id"))     unicode = g.get("id").getAsInt();
        else if (g.has("index"))  unicode = g.get("index").getAsInt() + 0xE000;
        if (unicode < 0) return;

        float advance = getFloat(g, "advance", 0) * fontSize;
        if (advance == 0) advance = getFloat(g, "xadvance", 0);

        float x = 0, y = 0, w = 0, h = 0, xOff = 0, yOff = 0;

        if (g.has("atlasBounds")) {
            JsonObject b = g.getAsJsonObject("atlasBounds");
            float left = getFloat(b, "left", 0), bottom = getFloat(b, "bottom", 0);
            float right = getFloat(b, "right", 0), top = getFloat(b, "top", 0);
            x = left;
            w = right - left;
            h = top - bottom;
            y = yOriginBottom ? atlasHeight - top : bottom;
        } else if (g.has("x")) {
            x = getFloat(g, "x", 0); y = getFloat(g, "y", 0);
            w = getFloat(g, "width", 0); h = getFloat(g, "height", 0);
        }

        if (g.has("planeBounds")) {
            JsonObject p = g.getAsJsonObject("planeBounds");
            xOff = getFloat(p, "left", 0) * fontSize;
            yOff = (ascender - getFloat(p, "top", 0)) * fontSize;
        } else {
            xOff = getFloat(g, "xoffset", 0);
            yOff = getFloat(g, "yoffset", 0);
        }

        glyphs.put(unicode, new Glyph(unicode, x, y, w, h, xOff, yOff, advance, atlasWidth, atlasHeight));
    }

    private float getFloat(JsonObject obj, String key, float def) {
        return obj.has(key) ? obj.get(key).getAsFloat() : def;
    }

    /**
     * Returns glyph for code point.
     * <p>
     * Возвращает глиф для кодовой точки.
     */
    public Glyph getGlyph(int codePoint)    { return glyphs.get(codePoint); }

    /**
     * Checks if glyph exists.
     * <p>
     * Проверяет наличие глифа.
     */
    public boolean hasGlyph(int codePoint)  { return glyphs.containsKey(codePoint); }

    /**
     * Returns texture identifier.
     * <p>
     * Возвращает идентификатор текстуры.
     */
    public Identifier getTextureId()        { return textureId; }

    /**
     * Returns font size.
     * <p>
     * Возвращает размер шрифта.
     */
    public float getFontSize()              { return fontSize; }

    /**
     * Returns line height.
     * <p>
     * Возвращает высоту строки.
     */
    public float getLineHeight()            { return lineHeight; }

    /**
     * Returns atlas width.
     * <p>
     * Возвращает ширину атласа.
     */
    public float getAtlasWidth()            { return atlasWidth; }

    /**
     * Returns atlas height.
     * <p>
     * Возвращает высоту атласа.
     */
    public float getAtlasHeight()           { return atlasHeight; }

    /**
     * Returns distance range.
     * <p>
     * Возвращает диапазон расстояния.
     */
    public float getDistanceRange()         { return distanceRange; }

    /**
     * Checks if loaded.
     * <p>
     * Проверяет загруженность.
     */
    public boolean isLoaded()               { return loaded.get(); }

    /**
     * Returns glyph count.
     * <p>
     * Возвращает количество глифов.
     */
    public int getGlyphCount()              { return glyphs.size(); }

    /**
     * Returns ascender.
     * <p>
     * Возвращает_ascender.
     */
    public float getAscender()              { return ascender; }
}

File: src/so/aporia/utils/user/render/font/FontAtlas.kt
/*
 * Copyright (c) 2025-2026 Aporia.cc Project
 * Distributed under the Aporia.cc Software License Agreement v1.0
 * See LICENSE and COPYRIGHT files in the project root for full text.
 */

package so.aporia.utils.user.render.font

import com.google.gson.JsonObject
import com.google.gson.JsonParser
import net.minecraft.client.Minecraft
import net.minecraft.resources.Identifier
import org.slf4j.LoggerFactory
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.HashMap
import java.util.concurrent.atomic.AtomicBoolean

class FontAtlas(private val jsonId: Identifier, private val textureId: Identifier) {

    private val glyphs: MutableMap<Int, Glyph> = HashMap()
    private val loaded = AtomicBoolean(false)

    private var atlasWidth: Float = 512f
    private var atlasHeight: Float = 512f
    private var fontSize: Float = 32f
    private var lineHeight: Float = 40f
    private var distanceRange: Float = 4f
    private var ascender: Float = 0.95f
    private var yOriginBottom: Boolean = false

    fun ensureLoaded() {
        if (loaded.get()) return
        synchronized(this) {
            if (loaded.get()) return
            doLoad()
        }
    }

    private fun doLoad() {
        try {
            val res = Minecraft.getInstance().resourceManager.getResource(jsonId)
            if (res.isEmpty) {
                LOGGER.warn("Font JSON not found in ResourceManager: {}, trying AssetManager...", jsonId)
                loadFromAssetManager()
                return
            }
            res.get().open().use { `is` ->
                InputStreamReader(`is`, StandardCharsets.UTF_8).use { reader ->
                    parseJson(JsonParser.parseReader(reader).asJsonObject)
                    loaded.set(true)
                    LOGGER.info("Loaded font: {} ({} glyphs)", jsonId, glyphs.size)
                }
            }
        } catch (e: Exception) {
            LOGGER.error("Failed to load font: {}", jsonId, e)
            loaded.set(true)
        }
    }

    private fun loadFromAssetManager() {
        try {
            val json = so.aporia.utils.assets.AssetManager.getResourceString(jsonId)

            if (json != null) {
                parseJson(JsonParser.parseString(json).asJsonObject)
                loaded.set(true)
                LOGGER.info("Loaded font from AssetManager: {} ({} glyphs)", jsonId, glyphs.size)
            } else {
                throw Exception("AssetManager returned null (file not found)")
            }
        } catch (e: Exception) {
            LOGGER.error("Failed to load font from AssetManager: {}", jsonId, e)
            loaded.set(true)
        }
    }

    private fun parseJson(root: JsonObject) {
        if (root.has("atlas")) {
            val atlas = root.getAsJsonObject("atlas")
            atlasWidth = getFloat(atlas, "width", 512f)
            atlasHeight = getFloat(atlas, "height", 512f)
            fontSize = getFloat(atlas, "size", 32f)
            distanceRange = getFloat(atlas, "distanceRange", 4f)
            if (atlas.has("yOrigin")) {
                yOriginBottom = atlas.get("yOrigin").asString.equals("bottom", ignoreCase = true)
            }
        }

        if (root.has("metrics")) {
            val metrics = root.getAsJsonObject("metrics")
            lineHeight = getFloat(metrics, "lineHeight", 1.2f) * fontSize
            ascender = getFloat(metrics, "ascender", 0.95f)
        }

        if (root.has("glyphs")) {
            for (elem in root.getAsJsonArray("glyphs")) {
                parseGlyph(elem.asJsonObject)
            }
        }
    }

    private fun parseGlyph(g: JsonObject) {
        var unicode = -1
        if (g.has("unicode")) unicode = g.get("unicode").asInt
        else if (g.has("char")) {
            val s = g.get("char").asString
            if (s.isNotEmpty()) unicode = s.codePointAt(0)
        } else if (g.has("id")) unicode = g.get("id").asInt
        else if (g.has("index")) unicode = g.get("index").asInt + 0xE000
        if (unicode < 0) return

        var advance = getFloat(g, "advance", 0f) * fontSize
        if (advance == 0f) advance = getFloat(g, "xadvance", 0f)

        var x = 0f; var y = 0f; var w = 0f; var h = 0f; var xOff = 0f; var yOff = 0f

        if (g.has("atlasBounds")) {
            val b = g.getAsJsonObject("atlasBounds")
            val left = getFloat(b, "left", 0f); val bottom = getFloat(b, "bottom", 0f)
            val right = getFloat(b, "right", 0f); val top = getFloat(b, "top", 0f)
            x = left
            w = right - left
            h = top - bottom
            y = if (yOriginBottom) atlasHeight - top else bottom
        } else if (g.has("x")) {
            x = getFloat(g, "x", 0f); y = getFloat(g, "y", 0f)
            w = getFloat(g, "width", 0f); h = getFloat(g, "height", 0f)
        }

        if (g.has("planeBounds")) {
            val p = g.getAsJsonObject("planeBounds")
            xOff = getFloat(p, "left", 0f) * fontSize
            yOff = (ascender - getFloat(p, "top", 0f)) * fontSize
        } else {
            xOff = getFloat(g, "xoffset", 0f)
            yOff = getFloat(g, "yoffset", 0f)
        }

        glyphs[unicode] = Glyph(unicode, x, y, w, h, xOff, yOff, advance, atlasWidth, atlasHeight)
    }

    private fun getFloat(obj: JsonObject, key: String, def: Float): Float {
        return if (obj.has(key)) obj.get(key).asFloat else def
    }

    fun getGlyph(codePoint: Int): Glyph? = glyphs[codePoint]

    fun hasGlyph(codePoint: Int): Boolean = glyphs.containsKey(codePoint)

    fun getTextureId(): Identifier = textureId

    fun getFontSize(): Float = fontSize

    fun getLineHeight(): Float = lineHeight

    fun getAtlasWidth(): Float = atlasWidth

    fun getAtlasHeight(): Float = atlasHeight

    fun getDistanceRange(): Float = distanceRange

    fun isLoaded(): Boolean = loaded.get()

    fun getGlyphCount(): Int = glyphs.size

    fun getAscender(): Float = ascender

    companion object {
        private val LOGGER = LoggerFactory.getLogger("aporia/Font")
    }
}
